Allow running behind proxy

mering opened the pull request:
See #210399 for the motivation.

Commits:
- Serve from `/` and basePath
- Honor `X-Forwarded-Prefix` header
- Honor `X-Forwarded-Port` header

Test procedure:
- Compile via `yarn compile`
- Run `scripts/code-server.sh`
- Run an nginx proxy with the following config:
    ```nginx
    server {
        listen 8080;
        listen [::]:8080;
    
        location /some/path/ {
            proxy_pass http://localhost:9888/;
            proxy_set_header Host $host;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection $http_connection;
            proxy_set_header X-Forwarded-Host $host;
            proxy_set_header X-Forwarded-Port $server_port;
            proxy_set_header X-Forwarded-Prefix "/some/path";
        }
    }
    ```
- Access VSCode on http://localhost:8080/some/path

Fixes #210399, #225706

mering committed:
Serve from '/' and basePath

jhgoebbert commented:
Any chance this small fix can be merged?

mering commented:
Unfortunately, there doesn't seem to be a way to test this change as described in https://github.com/microsoft/vscode/issues/210399#issuecomment-2058811322:

I tried getting a testing environment on `main` branch:
- Running `./scripts/code-cli.sh serve-web`
- First, I encountered #210454 when running in the Devcontainer.
- After manually installing Rust within the container, I only get `error error getting latest version: Updates are are not available: no configured quality`.

darkdragon-001 commented:
@aeschli @oakaigh @Tyriar Could you please take a look? Thanks!

dredwilliams reviewed (approved):
This is a simple change that appears to address the issue without having any other impacts

lost-in-dev-null reviewed (approved):
An easy fix for an annoying issue. I see no reason why not to merge.

aeschli commented:
When you start the server with `server-base-path`, that means that all the bits, including the main page are served from that path. If you want the server to be served from root ('/'), don't use that argument.

Why does the proxy forward the request to '/', but not '/server-base-path' ?

darkdragon-001 commented:
> Why does the proxy forward the request to '/', but not '/server-base-path' ?

This is just the way how most proxies are configured for simplicity and often end-users can't control this.

The server base path is just needed as a prefix for absolute links, not for actually serving files from that path.

darkdragon-001 commented:
Trying to explain why most proxies use the explained approach:

For which use-case would one use a different base path? When one wants to serve multiple instances or different services on a single host or domain. How do you distribute traffic in this case? You run a proxy and forward requests based on some rules to other services usually differentiated by a different local port on the host. As each service is only serving one instance, there is no use of using a different base path there but absolute links need to contain the correct prefix such that they work correctly.

dredwilliams commented:
We are trying to integrate code into an existing application that proxies requests into the code server with the base-path intact ... not our choice.  The expectation is that the code server will take the base-path into consideration when interpreting HTTP requests, and include that base-path on any referral links.  I thought that was what the base-path parameter was for ...

aeschli commented:
As mentioned, the idea of `server-base-path` is that the server serves the VS Code main page and all the resources the main requires from that path. With the PR it would no longer serve serves the VS Code main page from that path. That seems wrong.
At the very least it should serve the main page from / and from server-base-path.

aeschli commented:
Also, I'd like to learn more about that existing application that proxies requests into the code server

dredwilliams commented:
@aeschli - The the application is Open OnDemand, a web platform for accessing HPC resources.  It encapsulates an Apache web server and an NGINX application server to launch applications and drive a point-and-click interface for users.  It can launch applications that offer an HTML interface by proxying the connections through a dynamic path custom to each user and application.  That keeps all web traffic going through the same session with the user.

In this scenario, when the user wants to launch VS Code, we will build the unique path for that user's instance and provide that path via the base-path parameter.  We don't care that it can't respond to the default '/' path -- in fact, that's a desired impact when multiple users may be launching their own sessions on the same system.  Each user will have their own 'code serve-web' session with their unique base-path

mering commented:
> At the very least it should serve the main page from / and from server-base-path.

Serving from both paths sounds like a great idea to cover all use cases.
Unfortunately, it is almost impossible to create more complex patches without being able to test them (https://github.com/microsoft/vscode/pull/210455#issuecomment-2218715133). @aeschli could you prepare a PR for serving resources from both `/` and base path parameter if specified?

dredwilliams commented:
There is no need for the added complexity of responding to BOTH / and base-path ... if the base-path parameter is set, responding to '/' is not the desired behavior FOR THAT INSTANCE.

darkdragon-001 commented:
> There is no need for the added complexity of responding to BOTH / and base-path ... if the base-path parameter is set, responding to '/' is not the desired behavior FOR THAT INSTANCE.

As a lot of people explained including myself, responding only to '/' is the desired behavior for most proxy use cases.

Serving on both should fix all use cases.

mering commented:
I think I got it fixed to served from both, root and base path.

I basically remove the base path when the request comes in and add it for sending to the client.

mering committed:
Honor X-Forwarded-Prefix header

This allows determining the base path automatically if set correctly by the proxy.

mering committed:
Honor X-Forwarded-Port header

mering commented:
This approach allowed me to honor `X-Forwarded-Prefix` header correctly. When a proxy sets this correctly, paths are served correctly even without base path being explicitly set as argument.

mering commented:
> In this scenario, when the user wants to launch VS Code, we will build the unique path for that user's instance and provide that path via the base-path parameter. We don't care that it can't respond to the default '/' path -- in fact, that's a desired impact when multiple users may be launching their own sessions on the same system. Each user will have their own 'code serve-web' session with their unique base-path

@dredwilliams Each `code serve-web` session has it's own port, so there is no need to have a different base path there. Only the proxy needs separate prefixes to route the traffic to the correct port.

dredwilliams commented:
This looks great -- thanks for the effort y'all put into this!

lost-in-dev-null reviewed (approved):
Think this will make a lot of people happy! Thanks for the effort!

mering commented:
@aeschli could you please take another look? Thanks!

mering commented:
@chrmarti @aeschli friendly ping.

bitomukesh commented:
<div>
<h2>Changelist by Bito</h2>
    <p>This pull request implements the following key changes.</p>

<div>
<table width="100%">
<tbody>
  <tr>
    <th>Key Change</th>
    <th>Files Impacted</th>
  </tr>

<tr>
<td width="30%">
    <b>Feature Improvement</b> - Enhanced Proxy Support
</td>
<td width="70%">


<div>

-&nbsp;[`remoteExtensionHostAgentServer.ts`](https://github.com/microsoft/vscode/pull/210455/files#diff-a74809a9877a05cdf40ce0a260dae35b2b953b316a5688362f56a86309bde70b) - Added support for server base path and improved path handling
<div>

-&nbsp;[`webClientServer.ts`](https://github.com/microsoft/vscode/pull/210455/files#diff-86987149dfdb33da80e536afb11dbf1b18ab7137112ed164cfd5abc524439b72) - Implemented handling of forwarded headers and base path for proxy support
</div>
</td>
</tr>

</tbody>
</table>
</div>
</div>

mering commented:
@bpasero @sandy081 could anyone please take a look? Everything is tested and working, all issues have been resolved and this change is highly desired by many users.

nathanweeks commented:
This capability would be useful to many---perhaps most---computing centers that use Open OnDemand to provide web access to their computing resources.

ShabbirK commented:
This capability should also enable users to run VS Code on Kubeflow. It would be great to get a release with this PR merged soon.

vivodi commented:
Is this still relevant?

darkdragon-001 commented:
> Is this still relevant?

Yes, many people are eagerly waiting for this to get merged!

jhgoebbert commented:
+1 from me.
I've been following this PR from the beginning and hope it gets merged sooner rather than later. I wonder what exactly is holding it back.

aeschli commented:
I took this PR and added some (polishing) changes on top: See https://github.com/microsoft/vscode/pull/237411

Closing this PR in favor of https://github.com/microsoft/vscode/pull/237411
